base/reduce.jl

```diff
@@ -141,23 +141,26 @@ mr_empty(f, op::typeof(|), T) = false
 _mapreduce(f, op, A::AbstractArray) = _mapreduce(f, op, linearindexing(A), A)
 
 function _mapreduce{T}(f, op, ::LinearFast, A::AbstractArray{T})
-    n = Int(length(A))
-    if n == 0
-        return mr_empty(f, op, T)
-    elseif n == 1
-        return r_promote(op, f(A[1]))
-    elseif n < 16
-        fx1 = r_promote(op, f(A[1]))
-        fx2 = r_promote(op, f(A[2]))
-        s = op(fx1, fx2)
-        i = 2
-        while i < n
-            @inbounds Ai = A[i+=1]
-            s = op(s, f(Ai))
+    inds = linearindices(A)
+    n = length(inds)
+    @inbounds begin
```

Comment:
`@inbounds` doesn't propagate, even with inlining (unless so marked). We had `inbounds` previously for the `n<16` case, but it missed several `A[i]` accesses.